src/pip/_internal/index.py

```diff
@@ -767,68 +878,51 @@ def _log_skipped_link(self, link, reason):
     def _link_package_versions(self, link, search):
         """Return an InstallationCandidate or None"""
-        version = None
-        if link.egg_fragment:
-            egg_info = link.egg_fragment
-            ext = link.ext
-        else:
-            egg_info, ext = link.splitext()
-            if not ext:
-                self._log_skipped_link(link, 'not a file')
-                return
-            if ext not in SUPPORTED_EXTENSIONS:
-                self._log_skipped_link(
-                    link, 'unsupported archive format: %s' % ext,
-                )
-                return
-            if "binary" not in search.formats and ext == wheel_ext:
-                self._log_skipped_link(
-                    link, 'No binaries permitted for %s' % search.supplied,
-                )
-                return
-            if "macosx10" in link.path and ext == '.zip':
-                self._log_skipped_link(link, 'macosx10 one')
-                return
-            if ext == wheel_ext:
-                try:
-                    wheel = Wheel(link.filename)
-                except InvalidWheelFilename:
-                    self._log_skipped_link(link, 'invalid wheel filename')
-                    return
-                if canonicalize_name(wheel.name) != search.canonical:
-                    self._log_skipped_link(
-                        link, 'wrong project name (not %s)' % search.supplied)
-                    return
-
-                if not wheel.supported(self.valid_tags):
-                    self._log_skipped_link(
-                        link, 'it is not compatible with this Python')
-                    return
-
-                version = wheel.version
-
-        # This should be up by the search.ok_binary check, but see issue 2700.
-        if "source" not in search.formats and ext != wheel_ext:
+        try:
+            version = _parse_package_version(link, search, self.valid_tags)
+        except _LinkNotAFile:
+            self._log_skipped_link(link, 'not a file')
+            return
+        except _LinkFormatUnsupported:
+            self._log_skipped_link(
+                link, 'unsupported archive format: %s' % link.ext,
+            )
+            return
+        except _BinaryLinkNotPermitted:
+            self._log_skipped_link(
+                link, 'No binaries permitted for %s' % search.supplied,
+            )
+            return
+        except _MacOSXZip:
+            self._log_skipped_link(link, 'macosx10 one')
+            return
+        except InvalidWheelFilename:
+            self._log_skipped_link(link, 'invalid wheel filename')
+            return
+        except _IncorrectWheelName:
+            self._log_skipped_link(
+                link, 'wrong project name (not %s)' % search.supplied,
+            )
+            return
+        except _IncompatibleWheel:
+            self._log_skipped_link(
+                link, 'it is not compatible with this Python',
+            )
+            return
```

Comment:
Hmm. I can't help but think there's got to be a better way than all these `except-return`'s. One thing that doesn't seem right with this approach is that the caller is effectively adding to the exception information about what caused it, when the function raising the exception should be the one to do that -- because they were there when it was raised. In the case of `_IncorrectWheelName`, for example, what if that exception contained `link` and `search.supplied` as an attribute. Better yet, if the exception contained the whole message text, then all of these `try-except`'s could be reduced to a single `try-except` because it could just use the exception's message in each case. And you would still be eliminating the side effects you were worried about, which was the reason for breaking this out in the first place.